protocols/relay/src/v2/client.rs

```diff
@@ -269,16 +305,14 @@ impl NetworkBehaviour for Client {
                         event: Either::Left(handler::In::Reserve { to_listener }),
                     },
                     None => {
-                        let handler = handler::Prototype::new(
-                            self.local_peer_id,
-                            Some(handler::In::Reserve { to_listener }),
-                        );
+                        self.initial_events
+                            .insert(relay_peer_id, handler::In::Reserve { to_listener });
+
                         NetworkBehaviourAction::Dial {
                             opts: DialOpts::peer_id(relay_peer_id)
                                 .addresses(vec![relay_addr])
                                 .extend_addresses_through_behaviour()
```

Comment:
Note that some `NetworkBehaviour` might add a relayed address here.

The relay protocol should never be spoken on top of a relayed connection. We should never do nested relaying as otherwise one could build circular relayed connections that could be misused for a DOS attack. Thus returning a `dummy::ConnectionHandler` on relayed connections is critical.

That said, we might be establishing both a relayed and a direct connection to a given peer in parallel. How do we know which one the `initial_events` event was destined for?

Not sure how to prevent this. The solution you do above, namely to drop the `initial_events` event in case it turns out to be a relayed connection covers the case where there is only a single connection attempt in progress. What about the case where both (relayed and direct) are in progress?

Reply:
I only 1-to-1 ported the code that already existed, as far as I know, no functionality should currently change in this PR.